ykcompile/src/lib.rs

```diff
@@ -908,113 +508,216 @@ impl<TT> TraceCompiler<TT> {
+        match op {
             BinOp::Eq | BinOp::Ne | BinOp::Lt | BinOp::Le | BinOp::Gt | BinOp::Ge => {
-                self.c_condition(binop, op1, op2)
+                return self.c_condition(dest, &op, opnd1, opnd2);
             }
+            _ => {}
+        }
+
+        // We do this in three stages.
+        // 1) Copy the first operand into the temp register.
+        self.load_reg_iplace(*TEMP_REG, opnd1);
+
+        // 2) Add the second operand.
+        // FIXME replace with macro
+        let size = SIR.ty(&opnd1.ty()).size();
+        match op {
+            BinOp::Add => self.c_binop_add(dest, size, opnd2),
+            BinOp::Sub => self.c_binop_sub(dest, size, opnd2),
             _ => todo!(),
         }
-    }
 
-    fn c_condition(&mut self, binop: &BinOp, op1: &Operand, op2: &Operand) -> Location {
-        let (src1, ty) = match op1 {
-            Operand::Place(p) => self.place_to_location(&p, false),
-            Operand::Constant(Constant::Int(_)) => todo!(),
-            Operand::Constant(Constant::Bool(_b)) => todo!(),
-            Operand::Constant(c) => todo!("{}", c),
-        };
-        let src2 = match op2 {
-            Operand::Place(p) => self.c_place(&p),
-            Operand::Constant(Constant::Int(ci)) => self.c_constint(&ci),
-            Operand::Constant(Constant::Bool(_b)) => todo!(),
-            Operand::Constant(c) => todo!("{}", c),
-        };
+        // FIXME. Check for overflow here. Examine either CF or OF depending on signedness.
 
-        match (&src1, &src2) {
-            (Location::Register(reg1), Location::Register(reg2)) => match ty.size() {
-                1 => {
+        // 3) Move the result to where it is supposed to live.
+        // If it is a checked operation, then we have to build a (value, overflow?) tuple.
+        let mut dest_loc = self.iplace_to_location(dest);
+        self.store_raw(&dest_loc, &*TEMP_LOC, size);
+        if checked {
+            // Set overflow flag.
+            dynasm!(self.asm
+                ; mov Rq(*TEMP_REG), 0
+            );
+            let ro = dest_loc.unwrap_mem_mut();
+            ro.offs += i32::try_from(size).unwrap();
+            self.store_raw(&dest_loc, &*TEMP_LOC, 1);
+        }
+    }
+
+    fn c_binop_add(&mut self, dest: &IPlace, size: u64, opnd2: &IPlace) {
+        let src_loc = self.iplace_to_location(opnd2);
+        match src_loc {
+            Location::Register(r) => match size {
+                1 | 2 | 4 => todo!(),
+                8 => {
                     dynasm!(self.asm
-                        ; cmp Rb(reg1), Rb(reg2)
+                        ; add Rq(*TEMP_REG), Rq(r)
                     );
                 }
-                2 => {
-                    dynasm!(self.asm
-                        ; cmp Rw(reg1), Rw(reg2)
-                    );
+                _ => unreachable!(format!("{}", SIR.ty(&dest.ty()))),
```

Comment:
I think `unreachable!("{}", SIR.ty(...)` works?

Reply:
It does. Good catch.